### DISCUSSION TITLE
Transform existing class into a Document

### DISCUSSION THREAD

**Discussion by cperezabo** [mongodb/mongoid#5851] [Q&A] [answered]
Hi everyone!

Suppose I have the following class:

```ruby
class Dummy
  attr_accessor :name

  def initialize(name:)
    @name = name
  end

  def greet
    "Hi, I'm #{name}"
  end
end
```

At some point I want to convert it into a Document by sending it the expected messages:

```ruby
Dummy.include Mongoid::Document
Dummy.field :name
```

But it doesn't work, objects are not even saved and there is no error at all when sending the `save!` message.

I've also tried prepending the `Mongoid::Document` module but it breaks completely

Any idea on how to achieve this?

Thanks!

**Comment by jamis** [comment]
Hey @cperezabo  -- I'm afraid the solution isn't as simple as you're hoping. A big part of the problem here is that `Mongoid::Document` has its own constructor, and if it isn't called, things break down.

Not knowing exactly what you're wanting the end goal to look like, it's hard to know what to recommend. If at all possible, you'll want to make sure the `Dummy` constructor is compatible with the `Mongoid::Document` constructor, and make sure you call `super` if `Mongoid::Document` is present in the `ancestors` list.

The best approach is to ensure that your Mongoid model classes are _dedicated_ Mongoid model classes, designed around `Mongoid::Document` from the beginning. Then, you might add a method to your `Dummy` class for returning a Mongoid model, e.g.

```ruby
class Dummy
  class Model
    include Mongoid::Document
    field :name, type: String
  end

  attr_accessor :name

  def initialize(name:)
    @name = name
  end

  def to_model
    Model.new(name: name)
  end

  # ...
end
```

If there is any shared behavior between the Mongoid model and the pure-Ruby version, extract it to a module that you can include into each version.

Again, this might be severely over-engineering things, since I don't know what your requirements are, but hopefully I've given you some ideas to think about.

**Reply by cperezabo** [reply]
Hi, thank you for your time! I need to convert my class into a `Mongoid::Document` based on my environment. 

I have the following code
```ruby
# Base class
class Dummy
  attr_accessor :name

  def initialize(name:)
    @name = name
  end

  def present
    "Hi, I'm #{name}"
  end
end

# Conditional patch
class Dummy
  include Mongoid::Document
  field :name

  def initialize(*, **)
    super
  end
end

# Usage
puts Dummy.create!(name: "Pepe").present
puts Dummy.find_by.present
```

Output

```
Hi, I'm Pepe
Hi, I'm 
```

Fields are not being stored.

**Reply by jamis** [reply]
"Based on my environment." -- Does this mean, depending on whether a particular use-case calls for Mongoid or not?

If that's the case, you might try extracting the bits that differ between environments, to separate modules, and including them conditionally. Here's an example that demonstrates this:

```ruby
# The base class, with just the common, core functionality that is independent of
# whether the class is a Mongoid model or just a Plain Old Ruby Object (PORO).
class Dummy
  def present
    "Hi, I'm #{name}"
  end
end

if ARGV.include?('--mongoid')
  require 'mongoid'

  module WithMongoid
    extend ActiveSupport::Concern
    include Mongoid::Document

    included do
      field :name, type: String
    end
  end

  Dummy.include WithMongoid
else
  module WithoutMongoid
    attr_accessor :name

    def initialize(name:)
      self.name = name
    end
  end

  Dummy.include WithoutMongoid
end

d = Dummy.new(name: 'Jeff')
puts d.present
```

**Reply by cperezabo** [reply]
I always develop using plain objects with no relation to database. 
Doing that with data mappers like Doctrine or SQLAlchemy is easy, but with ActiveRecord-like mappers is not, but, we are in Ruby which has great meta-programming capabilities, so I am trying to hook up Mongoid to my objects so I can persist them transparently.
The problem with your proposal (thanks for that) is that the base Dummy class lacks of `initialize`, so it's not complete by itself.
I hope I'm being clear, English is not my main language.

**Reply by cperezabo** [reply]
And returning to https://github.com/mongodb/mongoid/discussions/5851#discussioncomment-10337845, what could be the reason why sending `field` after the `attr_accessor` is defined won't replace the accessor with the machinery to persist the field data correctly?

**Reply by cperezabo** [reply]
Ok, I've tried this

```ruby
class Dummy
  include Mongoid::Document

  remove_method :name
  remove_method :name=

  field :name

  def initialize(*, **)
    super
  end
end
```
And it worked! 
I think `field` should remove existing accessors to work safely or at least fail with an error. 
Saving nothing silently is not a good behavior. 
What do you think?

**Reply by jamis** [reply, marked as answer]
Yeah, interesting. I can't think of any reason why we would want `field` to silently fail in the presence of an existing accessor, but that doesn't mean there isn't a good reason for it. :) If you have time/inclination to explore that a bit in a PR, I'd be open to reviewing it. We would need to move a bit carefully since it changes existing behavior that (conceivably) someone could be relying on; it might need to be hidden behind a feature flag until the next major version of Mongoid.

**Reply by jamis** [reply]
As a quick follow-up, it should be noted that Ruby itself is silent when redefining an accessor:

```ruby
class A
  attr_accessor :name
  attr_accessor :name
end
```

However, running it with `-w` to enable warnings does show a warning in that case. Mongoid should probably, at the very minimum, follow that precedent.
